clientele/framework/client.py

```diff
@@ -0,0 +1,406 @@
+                transport=transport,
+                cookies=cookies,
+            )
+        self.config = config
+        self._client: httpx.Client | None = None
+        self._async_client: httpx.AsyncClient | None = None
+
+    def __enter__(self) -> "Client":
+        self.open()
+        return self
+
+    def __exit__(self, exc_type, exc, tb) -> None:  # type: ignore[override]
+        self.close()
+
+    async def __aenter__(self) -> "Client":
+        await self.aopen()
+        return self
+
+    async def __aexit__(self, exc_type, exc, tb) -> None:  # type: ignore[override]
+        await self.aclose()
+
+    def open(self) -> None:
+        """Instantiate a persistent ``httpx.Client``."""
+
+        if self._client is None:
+            self._client = self._build_client()
+
+    async def aopen(self) -> None:
+        """Instantiate a persistent ``httpx.AsyncClient``."""
+
+        if self._async_client is None:
+            self._async_client = self._build_async_client()
+
+    def close(self) -> None:
+        """Close the persistent client if it exists."""
+
+        if self._client is not None:
+            self._client.close()
+            self._client = None
+
+    async def aclose(self) -> None:
+        """Close the persistent async client if it exists."""
+
+        if self._async_client is not None:
+            await self._async_client.aclose()
+            self._async_client = None
+
+    def get(self, path: str) -> Callable[[_F], _F]:
+        return self._create_decorator("GET", path)
+
+    def post(self, path: str) -> Callable[[_F], _F]:
+        return self._create_decorator("POST", path)
+
+    def put(self, path: str) -> Callable[[_F], _F]:
+        return self._create_decorator("PUT", path)
+
+    def patch(self, path: str) -> Callable[[_F], _F]:
+        return self._create_decorator("PATCH", path)
+
+    def delete(self, path: str) -> Callable[[_F], _F]:
+        return self._create_decorator("DELETE", path)
+
+    def _create_decorator(self, method: str, path: str) -> Callable[[Callable[..., Any]], Callable[..., Any]]:
+        def decorator(func: Callable[..., Any]) -> Callable[..., Any]:
+            context = _build_request_context(method, path, func)
+
+            if inspect.iscoroutinefunction(func):
+
+                @wraps(func)
+                async def async_wrapper(*args: Any, **kwargs: Any) -> Any:
+                    return await self._execute_async(context, args, kwargs)
+
+                return async_wrapper
+
+            @wraps(func)
+            def wrapper(*args: Any, **kwargs: Any) -> Any:
+                return self._execute_sync(context, args, kwargs)
+
+            return wrapper
+
+        return decorator
+
+    def _build_client(self) -> httpx.Client:
+        return httpx.Client(**self.config.httpx_client_options())
+
+    def _build_async_client(self) -> httpx.AsyncClient:
+        return httpx.AsyncClient(**self.config.httpx_client_options())
+
+    @contextmanager
```

Comment:
I see support for a context manager but no documentation for it nor any tests. Do we need it?